NETCORE/src/Shared/Implementation/TelemetryConfigurationOptionsSetup.cs

```diff
@@ -194,23 +207,31 @@ public void Configure(TelemetryConfiguration configuration)
                     }
 
                     // AppServicesHeartbeatTelemetryModule
-                    if (module is AppServicesHeartbeatTelemetryModule)
+                    if (module is AppServicesHeartbeatTelemetryModule appServicesHeartbeatTelemetryModule)
                     {
                         if (!this.applicationInsightsServiceOptions.EnableAppServicesHeartbeatTelemetryModule)
                         {
                             DisposeIfDisposable(module);
                             continue;
                         }
+                        else
+                        {
+                            HeartbeatHelper.SetHeartbeatPropertyManager(appServicesHeartbeatTelemetryModule, heartbeatPropertyManager);
```

Comment:
I'd use reflection only if every other option is exhausted. Did we try alternate options instead of reflection hack?
The Asp.Net Core/Worker knows about heartbeat, appservice module etc - so can we modify the appservice module to take the heartmanager via constructor?

Reply:
I didn't investigate that because of a few unknowns.
- DI can have conflicts when adding overload constructors. 
- We're registering DiagnosticTelemetryModule as ITelemetry, but would need to use IHeartbeatPropertyManager in a constructor.

Reply:
So I looked into this and I think we can make it work, but I want to review this with you before I make some major changes.

I'll have to add a new constructor to both AzureInstanceMetadataTelemetryModule and AppServiceHeartbeatTelemetryModule.
```
        public AzureInstanceMetadataTelemetryModule(IHeartbeatPropertyManager heartbeatPropertyManager)
        {
            this.HeartbeatPropertyManager = heartbeatPropertyManager;
        }
```

but to make this work, we HAVE to register DiagnosticTelemetryModule with the other interface.
```
        services.AddSingleton<IHeartbeatPropertyManager, DiagnosticsTelemetryModule>();
```

This is SIGNIFICANTLY cleaner than the reflection approach.
But I don't know how this will affect customers who register an instance of the Module.
We've been instructing users to do this:
```
        services.AddSingleton<ITelemetryModule>(new DiagnosticsTelemetryModule());
```

I don't know how to make this work with that, it'll take some extra testing.
What do you think?

Reply:
This article suggests some workarounds.
https://andrewlock.net/how-to-register-a-service-with-multiple-interfaces-for-in-asp-net-core-di/

I'll test these tomorrow
```
    services.AddSingleton<Foo>(); // We must explicitly register Foo
    services.AddSingleton<IFoo>(x => x.GetRequiredService<Foo>()); // Forward requests to Foo
    services.AddSingleton<IBar>(x => x.GetRequiredService<Foo>()); // Forward requests to Foo
```